<span style="font-family: Arial; font-weight:bold;font-size:2em;color:#0ab6fa">Hands on Linear Programming with Gurobi

**Phillipe Vilaça Gomes** <br>
**Institute for Research in Technology IIT, Comillas Pontifical University** <br>
**Madrid, Spain, 2024.** <br>
- Phillipe Vilaça, phillipe.v@comillas.edu

**Note**: <br>
This educational material on "Hands-on Linear Programming with Gurobi" is derived from the "Interpretable Optimization" project. This project was conducted by: <br>
- Dra Sara Lumbreras (https://www.iit.comillas.edu/personas/slumbreras);
- Dr Javier García González (https://www.iit.comillas.edu/personas/javiergg); and 
- Dr Phillipe Vilaça Gomes (https://www.iit.comillas.edu/personas/phillipe.v).

**Prerequisite Knowledge Notice:** <br>
This educational material on "Hands-on Linear Programming with Gurobi" is designed for students who possess at least an intermediate level of understanding in key areas of Linear Programming (LP). Before delving into this content, it is recommended that students are familiar with the following concepts:

- **LP Formulation:** Understanding the basic mathematical formulation of linear programming problems.
- **LP Feasible Region and Optimality:** Knowledge of what constitutes a feasible region in LP and how optimality is determined within this context.
- **LP Sensitivity Analysis:** Insight into how changes in the parameters of a linear program affect its solution.
- **LP Duality:** A grasp of the principles of duality in linear programming, including the ability to understand and interpret dual problems.

This foundational knowledge is crucial for a comprehensive understanding of the material presented, as it builds upon these core concepts, particularly in the application of the Gurobi solver for LP problems.



## Table of Content
**[1. Introduction to Gurobi](#M1)** <br>
**[2. Installing Gurobi](#M2)** <br>
**[3. Importing Gurobi and other libraries](#M3)** <br>
**[4. Detailed LP Problem: A real world example](#M4)** <br>
**[5. Detailed LP Problem: Mathematical Formulation](#M5)** <br>
**[6. Detailed LP Problem: Gurobi Model](#M6)** <br>
**[7. Key Gurobi commands](#M7)** <br>

<a id="M1"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">1. Introduction to Gurobi

- Gurobi is a powerful optimization solver that specializes in linear programming (LP), mixed-integer programming (MIP), and other types of mathematical programming. It's widely used in both academia and industry due to its efficiency and robustness in handling complex optimization problems. Gurobi is known for its high performance and ability to provide accurate solutions rapidly.

- In this material, we will explore the basic functionalities of Gurobi and demonstrate how to set up and solve linear programming problems using this solver. This hands-on approach will provide practical insights into utilizing Gurobi for real-world optimization challenges.


<a id="M2"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">2. Installing Gurobi

In [46]:
# Uncomment and run this cell if you need to install Gurobi
# !pip install gurobipy

<a id="M3"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">3 Importing Gurobi and other libraries

In [5]:
import gurobipy as gp
from gurobipy import GRB
import os
import numpy as np
import warnings

# Suppress DeprecationWarning
warnings.filterwarnings('ignore', category=DeprecationWarning)

<a id="M4"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">4 Detailed LP Problem: A real world example

### Scenario

A factory manufactures two types of products: Product A and Product B. The factory operates with two machines, Machine 1 and Machine 2. The goal is to maximize the factory's profits under certain production and machine usage constraints.

### Products and Profit

- **Product A** brings a profit of \$50 per unit.
- **Product B** brings a profit of \$40 per unit.

### Production Time Constraints

- **Machine 1** has a maximum operational time of 12 hours per day.
   - It takes 1 hour to produce one unit of Product A.
   - It takes 0.5 hours to produce one unit of Product B.

- **Machine 2** has a maximum operational time of 10 hours per day.
   - It takes 0.5 hours to produce one unit of Product A.
   - It takes 1 hour to produce one unit of Product B.

### Market Demand Constraint

- The market can absorb a maximum of 20 units of Product A and 25 units of Product B per day.

### Objective

Determine the optimal number of units of Product A and Product B to produce daily to maximize profits, considering machine time and market demand constraints.

<a id="M5"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">5 Detailed LP Problem: Mathematical Formulation

### Decision Variables
- \( $x_1$ \): Number of units of Product A produced per day.
- \( $x_2$ \): Number of units of Product B produced per day.

### Objective Function
- Maximize total profit: \( Z = 50$x_1$ + 40$x_2$ \)

### Constraints
1. Machine 1 Time: \( $x_1$ + 0.5$x_2$ $\leq$ 12 \)
2. Machine 2 Time: \( 0.5$x_1$ + $x_2$ $\leq$ 10 \)
3. Market Demand for Product A: \( $x_1$ $\leq$ 20 \)
4. Market Demand for Product B: \( $x_2$ $\leq$ 25 \)
5. Non-negativity: \( $x_1$, $x_2$ $\geq$ 0 \)


<a id="M6"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">6 Detailed LP Problem: Gurobi Model

In [6]:
# Define the model
model = gp.Model("FactoryOptimization")

Set parameter Username
Academic license - for non-commercial use only - expires 2024-12-04


In [7]:
# Add Decision Variables
x1 = model.addVar(vtype=GRB.CONTINUOUS, name="x1", lb=0)
x2 = model.addVar(vtype=GRB.CONTINUOUS, name="x2", lb=0)

In [8]:
# Set the Objective Function
model.setObjective(50 * x1 + 40 * x2, GRB.MAXIMIZE)

In [9]:
# Add Constraints

# Machine 1 time constraint
model.addConstr(x1 + 0.5 * x2 <= 12, "Machine1")

# Machine 2 time constraint
model.addConstr(0.5 * x1 + x2 <= 10, "Machine2")

# Market demand constraints
model.addConstr(x1 <= 20, "DemandA")
model.addConstr(x2 <= 25, "DemandB")


<gurobi.Constr *Awaiting Model Update*>

In [10]:
# Solve the Model
model.optimize()

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (win64 - Windows 11+.0 (22631.2))

CPU model: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 4 rows, 2 columns and 6 nonzeros
Model fingerprint: 0xbeafecd7
Coefficient statistics:
  Matrix range     [5e-01, 1e+00]
  Objective range  [4e+01, 5e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+01, 3e+01]
Presolve removed 2 rows and 0 columns
Presolve time: 0.01s
Presolved: 2 rows, 2 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    9.6000000e+02   1.399450e+01   0.000000e+00      0s
       2    6.8000000e+02   0.000000e+00   0.000000e+00      0s

Solved in 2 iterations and 0.02 seconds (0.00 work units)
Optimal objective  6.800000000e+02


In [11]:
# Print the solution
if model.status == GRB.OPTIMAL:
    print(f"Optimal number of Product A to produce: {x1.X}")
    print(f"Optimal number of Product B to produce: {x2.X}")
    print(f"Maximum Profit: ${model.ObjVal}")

Optimal number of Product A to produce: 9.333333333333334
Optimal number of Product B to produce: 5.333333333333333
Maximum Profit: $680.0


In [12]:
# Check if the optimization was successful
if model.status == GRB.OPTIMAL:
    # Retrieve the optimal values
    x1_optimal = x1.X
    x2_optimal = x2.X

    # Get the constraints from the model
    constraints = model.getConstrs()

    # Calculate and print LHS vs RHS for each constraint
    print("Constraint Comparisons (LHS vs RHS):")
    for constr in constraints:
        # Get the name of the constraint
        constr_name = constr.ConstrName

        # Get the RHS of the constraint
        rhs = constr.RHS

        # Calculate the LHS of the constraint
        # This requires summing the products of the variables' coefficients and their optimal values
        lhs = sum(model.getCoeff(constr, var) * var.X for var in model.getVars())

        # Print LHS and RHS
        print(f"{constr_name}: LHS = {lhs}, RHS = {rhs}")
else:
    print("Optimization was unsuccessful. Status code:", model.status)

Constraint Comparisons (LHS vs RHS):
Machine1: LHS = 12.0, RHS = 12.0
Machine2: LHS = 10.0, RHS = 10.0
DemandA: LHS = 9.333333333333334, RHS = 20.0
DemandB: LHS = 5.333333333333333, RHS = 25.0


**Note: Variable Types and Problem Complexity**

In our current model, the decision variables representing the quantities of Product A (`x1`) and Product B (`x2`) are treated as continuous variables. This means that our model may suggest fractional production quantities, such as producing 9.33 units of Product A.

However, in a real-world manufacturing context, it often makes more sense to have these quantities as integers since you cannot produce a fraction of a product. In such cases, the decision variables would be defined as integer variables, transforming our problem into a Mixed-Integer Linear Programming (MILP) problem.

MILP problems are generally more complex to solve than their continuous counterparts due to the added complexity of integer constraints. For the sake of simplicity and focus on the basic concepts of Linear Programming, we have not included integer constraints in this model. Nevertheless, it's important to be aware of this distinction and its implications in practical applications.


<a id="M7"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.8em;color:#6E1B1B">7 Key Gurobi commands

<a id="M7.1"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.0em;color:#6E1B1B">7.1 Reading a mps or a lp file

In [71]:
# Get the current working directory
current_working_directory = os.getcwd()

In [72]:
# Name of the model to read
model_name_mps = 'TRNSPORT.mps'  # MPS format

# Alternatively, you can use the LP format of the same model:
model_name_lp = 'TRNSPORT.lp'

**Note: File Formats** <br>
Whether you use the MPS or LP version, the Gurobi model created will be the same.
The MPS format (Mathematical Programming System) and LP format (Linear Programming format) are just different ways of encoding the same optimization problem. The MPS format is older and more compact, while the LP format is more human-readable.

In [73]:
# Read the model
model = gp.read(model_name_mps)
model_just_to_compare = gp.read(model_name_lp)

Read MPS format model from file TRNSPORT.mps
Reading time = 0.02 seconds
Convert: 6 rows, 7 columns, 19 nonzeros
Read LP format model from file TRNSPORT.lp
Reading time = 0.02 seconds
_obj: 5 rows, 7 columns, 12 nonzeros


In [74]:
# print the model (build from mps file)
model.display()

Minimize
  x7
Subject To
  e1: -0.225 x1 + -0.153 x2 + -0.162 x3 + -0.225 x4 + -0.162 x5 + -0.126 x6 + x7 = 0
  e2: x1 + x2 + x3 <= 350
  e3: x4 + x5 + x6 <= 600
  e4: x1 + x4 >= 325
  e5: x2 + x5 >= 300
  e6: x3 + x6 >= 275
Bounds
  x7 free


In [75]:
# print the model (build from lp file)
model_just_to_compare.display()

Minimize
0.225 x(seattle,newmyork)#0 + 0.153 x(seattle,chicago)#1 + 0.162 x(seattle,topeka)#2
+ 0.225 x(sanmdiego,newmyork)#3 + 0.162 x(sanmdiego,chicago)#4
+ 0.126 x(sanmdiego,topeka)#5 + constobj#6
Subject To
supply(seattle)#0: x(seattle,newmyork)#0 + x(seattle,chicago)#1 + x(seattle,topeka)#2
 <= 350
supply(sanmdiego)#1: x(sanmdiego,newmyork)#3 + x(sanmdiego,chicago)#4 +
 x(sanmdiego,topeka)#5 <= 600
  demand(newmyork)#2: x(seattle,newmyork)#0 + x(sanmdiego,newmyork)#3 >= 325
  demand(chicago)#3: x(seattle,chicago)#1 + x(sanmdiego,chicago)#4 >= 300
  demand(topeka)#4: x(seattle,topeka)#2 + x(sanmdiego,topeka)#5 >= 275
Bounds
  constobj#6 = 0


<a id="M7.2"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.0em;color:#6E1B1B">7.2 Extract key matrices and vectors from a Gurobi model 

## Generic Linear Programming Model Formulation

We define a generic linear programming (LP) problem in the following form:

**Objective:**
Minimize \( Z = $c^T x$ \)

**Subject to:**

\[ Ax $\geq$ b \] <br>
\[ lb $\leq$ x $\leq$ ub \]

In which:
- $b = [b_1, b_2, ..., b_n]$ is the right-hand side vector of the constraints.
- $c = [c_1, c_2, ..., c_n]$  is the coefficients vector for the objective function.
- $A = \begin{bmatrix}
  a_{11} & a_{12} & \cdots & a_{1n} \\
  a_{21} & a_{22} & \cdots & a_{2n} \\
  \vdots & \vdots & \ddots & \vdots \\
  a_{m1} & a_{m2} & \cdots & a_{mn}
\end{bmatrix}$ is the coefficient matrix for the constraints.
- $lb = [lb_1, lb_2, ..., lb_n]$ and $ub = [ub_1, ub_2, ..., ub_n]$ are the lower and upper bounds vectors for the decision variables $x$.


In [76]:
# Access constraint matrix A
A = model.getA()
print(A.A)

[[-0.225 -0.153 -0.162 -0.225 -0.162 -0.126  1.   ]
 [ 1.     1.     1.     0.     0.     0.     0.   ]
 [ 0.     0.     0.     1.     1.     1.     0.   ]
 [ 1.     0.     0.     1.     0.     0.     0.   ]
 [ 0.     1.     0.     0.     1.     0.     0.   ]
 [ 0.     0.     1.     0.     0.     1.     0.   ]]


In [77]:
# Access right-hand side (RHS) vector b
b = model.getAttr('RHS', model.getConstrs())
print(b)

[0.0, 350.0, 600.0, 325.0, 300.0, 275.0]


In [78]:
# Access objective function coefficients c
c = model.getAttr('Obj', model.getVars())
print(c)

[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]


In [79]:
# Access lower bounds (lb) 
lb = np.array([var.LB for var in model.getVars()])
print(lb)

[  0.   0.   0.   0.   0.   0. -inf]


In [80]:
# Access upper bounds (ub)
ub = np.array([var.UB for var in model.getVars()])
print(ub)

[inf inf inf inf inf inf inf]


In [81]:
# Access the sense of optimization
of_sense = model.ModelSense
print(of_sense)

1


In [82]:
# Access the sense of each constraint
cons_senses = [constr.Sense for constr in model.getConstrs()]
print(cons_senses)

['=', '<', '<', '>', '>', '>']


In [83]:
# Saving the model in lp format
model.write("model.lp")

In [84]:
# Reading the lp model
with open("model.lp", 'r') as file:
    model_lp = file.read()

print(model_lp)

\ Model Convert
\ LP format - for model browsing. Use MPS format to capture full model detail.
Minimize
  x7
Subject To
 e1: - 0.225 x1 - 0.153 x2 - 0.162 x3 - 0.225 x4 - 0.162 x5 - 0.126 x6 + x7
   = 0
 e2: x1 + x2 + x3 <= 350
 e3: x4 + x5 + x6 <= 600
 e4: x1 + x4 >= 325
 e5: x2 + x5 >= 300
 e6: x3 + x6 >= 275
Bounds
 x7 free
End



<a id="M7.3"> </a>
<span style="font-family: Arial; font-weight:bold;font-size:1.0em;color:#6E1B1B">7.3 Gurobi key commands

In [85]:
# Solving the model
model.optimize()

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (win64 - Windows 11+.0 (22631.2))

CPU model: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 6 rows, 7 columns and 19 nonzeros
Model fingerprint: 0x9e893693
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+02, 6e+02]
Presolve removed 1 rows and 1 columns
Presolve time: 0.01s
Presolved: 5 rows, 6 columns, 12 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   1.125000e+02   0.000000e+00      0s
       4    1.5367500e+02   0.000000e+00   0.000000e+00      0s

Solved in 4 iterations and 0.03 seconds (0.00 work units)
Optimal objective  1.536750000e+02


In [86]:
# Getting the values of the decision variables as a dictionary
{var.varName: var.x for var in model.getVars()}

{'x1': 50.0,
 'x2': 300.0,
 'x3': 0.0,
 'x4': 275.0,
 'x5': 0.0,
 'x6': 275.0,
 'x7': 153.675}

In [87]:
# Getting the values of the decision variables as a list
[var.x for var in model.getVars()]

[50.0, 300.0, 0.0, 275.0, 0.0, 275.0, 153.675]

In [88]:
# Getting the values of the decision variables as a scalar
var_index = 0
model.getVars()[var_index].x

50.0

In [89]:
# Check the optimization status
model.Status

2

**Note:**

The possible outcomes for model.Status are:
- 1: LOADED (Model is loaded, but no solution information is available)
- 2: OPTIMAL (Model was solved to optimality)
- 3: INFEASIBLE (Model was proven to be infeasible)
- 4: INF_OR_UNBD (Model was proven to be infeasible or unbounded)
- 5: UNBOUNDED (Model was proven to be unbounded)
- 6: CUTOFF (Optimal objective for model was worse than the specified cutoff)
- 7: ITERATION_LIMIT (Optimization terminated because the total number of simplex iterations performed exceeded the specified limit)
- 8: NODE_LIMIT (Optimization terminated because the total number of branch-and-cut nodes explored exceeded the specified limit)
- 9: TIME_LIMIT (Optimization terminated because the time expended exceeded the specified limit)
- 10: SOLUTION_LIMIT (Optimization terminated because the number of solutions found reached the specified limit)
- 11: INTERRUPTED (Optimization was terminated by the user)
- 12: NUMERIC (Optimization was terminated due to unrecoverable numerical difficulties)
- 13: SUBOPTIMAL (Unable to satisfy optimality tolerances; a sub-optimal solution is available)
- 14: INPROGRESS (An asynchronous optimization call was made, but the associated optimization run is not yet complete)
- 15: USER_OBJ_LIMIT (User specified an objective limit (a bound on either the best objective or the best bound), and that limit has been reached)

In [90]:
# Optimal solution value
model.ObjVal

153.675

**Note: Reduced cost**

- Reduced Cost (RC) in linear programming indicates how much the objective function coefficient of a non-basic 
variable (at its current value) would have to improve (increase for a minimization problem or decrease for a  maximization problem) before that variable could take a positive value in the solution. 
- In simple terms, it represents the amount by which the objective function's value per unit of the variable would need to improve for the variable to be included in the solution. For basic variables (variables that are part of the optimal solution), the Reduced Cost is zero.

In [91]:
# Getting the reduced costs of all decision variables as a dictionary
{var.varName: var.RC for var in model.getVars()}

{'x1': 0.0,
 'x2': 0.0,
 'x3': 0.036000000000000004,
 'x4': 0.0,
 'x5': 0.009000000000000008,
 'x6': 0.0,
 'x7': 0.0}

In [92]:
# Getting the reduced costs of all decision variables as a list
[var.RC for var in model.getVars()]

[0.0, 0.0, 0.036000000000000004, 0.0, 0.009000000000000008, 0.0, 0.0]

In [93]:
# Getting the reduced costs of all decision variables as a scalar
var_index = 2
model.getVars()[var_index].RC

0.036000000000000004

In [94]:
# getting the coefficient of the cost vector
var_index = 6
model.getVars()[var_index].Obj

1.0

**Note: Shadow Prices**


- The dual value indicates how much the objective value would change if the right-hand side of this constraint is increased by one unit

In [95]:
# Getting the dual values (shadow prices) of all constraints as a dictionary
{constr.ConstrName: constr.Pi for constr in model.getConstrs()}

{'e1': 1.0, 'e2': 0.0, 'e3': 0.0, 'e4': 0.225, 'e5': 0.153, 'e6': 0.126}

In [96]:
# Getting the dual values (shadow prices) of all constraints as a list
[constr.Pi for constr in model.getConstrs()]

[1.0, 0.0, 0.0, 0.225, 0.153, 0.126]

In [97]:
# Getting the dual values (shadow prices) of a constraints as a scalar
cons_index = 2
model.getConstrs()[cons_index].PI

0.0

**Note: Slacks**

- In linear programming, the slack of a constraint measures how far a constraint is from being binding (or tight) in the optimal solution. In other words, it indicates the amount by which the left-hand side (LHS) of a constraint can still change without changing the optimal solution.
- A slack of zero means the constraint is exactly met (binding).

In [98]:
# Getting the slack of all constraints as a dictionary
{constr.ConstrName: constr.Slack for constr in model.getConstrs()}

{'e1': 0.0, 'e2': 0.0, 'e3': 50.0, 'e4': 0.0, 'e5': 0.0, 'e6': 0.0}

In [99]:
# Getting the slack of all constraints as a dictionary
[constr.Slack for constr in model.getConstrs()]

[0.0, 0.0, 50.0, 0.0, 0.0, 0.0]

In [100]:
# Getting the slack of a constraints as a scalar
cons_index = 2
model.getConstrs()[cons_index].Slack

50.0

**Note: RHS Sensitivity Analysis**

- SARHSLow and SARHSUp provide the lower and upper limits for the RHS of the constraint within which the current optimal basis remains optimal.

In [60]:
# sensitivity analysis right-hand side low (SARHSLow)
{constr.ConstrName: constr.SARHSLow for constr in model.getConstrs()}

{'e1': -inf, 'e2': 300.0, 'e3': 550.0, 'e4': 50.0, 'e5': 25.0, 'e6': -0.0}

In [61]:
# sensitivity analysis right-hand side up (SARHSUp)
{constr.ConstrName: constr.SARHSUp for constr in model.getConstrs()}

{'e1': inf, 'e2': 625.0, 'e3': inf, 'e4': 375.0, 'e5': 350.0, 'e6': 325.0}

**Note: Intepretation of results**

| Constraint | Dual Value (Shadow Price) | Slack | SARHSLow | SARHSUp |
|------------|---------------------------|-------|----------|---------|
| e1         | 1.0                       | 0.0   | -inf     | inf     |
| e2         | 0.0                       | 0.0   | 300.0    | 625.0   |
| e3         | 0.0                       | 50.0  | 550.0    | inf     |
| e4         | 0.225                     | 0.0   | 50.0     | 375.0   |
| e5         | 0.153                     | 0.0   | 25.0     | 350.0   |
| e6         | 0.126                     | 0.0   | -0.0     | 325.0   |


- **Constraint e1**: 
  - Dual Value: 1.0, indicating that for each unit increase in the RHS, the objective function will improve by 1 unit.
  - Slack: 0.0, implying the constraint is binding (tight) at the optimal solution.
  - SARHSLow: -inf and SARHSUp: inf, suggesting that there is no limit to how much the RHS can be decreased or increased without changing the optimal basis.
- **Constraint e2**: 
  - Dual Value: 0.0, suggesting no improvement in the objective function with a unit increase in the RHS.
  - Slack: 0.0, indicating the constraint is binding.
  - SARHSLow: 300.0 and SARHSUp: 625.0, meaning the RHS can be decreased to 300 or increased to 625 without affecting the optimal solution.
- **Constraint e3**: 
  - Dual Value: 0.0, meaning no direct impact on the objective function from changes in the RHS.
  - Slack: 50.0, indicating that this constraint is not binding and has some flexibility.
  - SARHSLow: 550.0 and SARHSUp: inf, implying that decreasing RHS below 550 could change the optimal basis, but increasing has no upper bound.
- **Constraint e4**: 
  - **Dual Value**: 0.225, indicating that for each unit increase in the RHS, the objective function will improve by 0.225 units.
  - **Slack**: 0.0, showing that the constraint is binding at the optimal solution.
  - **SARHSLow**: 50.0 and **SARHSUp**: 375.0, meaning the RHS can vary between 50 and 375 without altering the optimal solution.

- **Constraint e5**: 
  - **Dual Value**: 0.153, suggesting that a unit increase in the RHS will result in an improvement of 0.153 units in the objective function.
  - **Slack**: 0.0, indicating that the constraint is tight and active in determining the optimal solution.
  - **SARHSLow**: 25.0 and **SARHSUp**: 350.0, showing that the feasible range for the RHS to change without impacting the optimality is between 25 and 350.

- **Constraint e6**: 
  - **Dual Value**: 0.126, signifying that the objective function value will increase by 0.126 for every unit increase in the RHS.
  - **Slack**: 0.0, implying that the constraint is binding.
  - **SARHSLow**: -0.0 and **SARHSUp**: 325.0, indicating that the RHS cannot decrease from its current value but can increase up to 325 without affecting the current optimal solution.